src/compiler/checker.ts

```diff
@@ -13667,21 +13669,15 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return instantiateType(instantiable, createTypeMapper([type.indexType, type.objectType], [getNumberLiteralType(0), createTupleType([replacement])]));
     }
 
-    // If the original mapped type had an intersection constraint we extract its components,
-    // and we make an attempt to do so even if the intersection has been reduced to a union.
-    // This entire process allows us to possibly retrieve the filtering type literals.
-    // e.g. { [K in keyof U & ("a" | "b") ] } -> "a" | "b"
-    function getLimitedConstraint(type: ReverseMappedType) {
+    // If the original mapped type had an union/intersection constraint
+    // there is a chance that it includes an intersection that could limit what members are allowed
+    function getReverseMappedTypeMembersLimitingConstraint(type: ReverseMappedType) {
         const constraint = getConstraintTypeFromMappedType(type.mappedType);
-        if (!(constraint.flags & TypeFlags.Union || constraint.flags & TypeFlags.Intersection)) {
-            return;
-        }
-        const origin = (constraint.flags & TypeFlags.Union) ? (constraint as UnionType).origin : (constraint as IntersectionType);
-        if (!origin || !(origin.flags & TypeFlags.Intersection)) {
+        if (constraint === type.constraintType) {
             return;
         }
-        const limitedConstraint = getIntersectionType((origin as IntersectionType).types.filter(t => t !== type.constraintType));
-        return limitedConstraint !== neverType ? limitedConstraint : undefined;
+        const mapper = appendTypeMapping(type.mappedType.mapper, type.constraintType.type, keyofConstraintObjectType);

```

Comment:
@weswigham would you mind taking a look at this PR? It's a fix to the recently-ish merged https://github.com/microsoft/TypeScript/pull/55811 

I have a concern that using `keyofConstraintObjectType` here is not that great since it might change the meaning of `T[K]`. In reality, I'd like to just map `keyof T`... but type mappings only work on type parameters.

Reply:
Hm, swapping `keyof Whatever` out for `string | number | symbol` and getting the resulting type is the goal here, right? That definitely aughta result in the useful bound. It's definitely possible `T[K]` if it were used in a constraint could return weird stuff here by doing an instantiation in this way (which could be actually done because of some conditional type weirdness).... hmmm... Maybe instantiate with `T` with `T & keyofConstraintObjectType` so specific-key indexing still turns up the specific member types, but `keyof` returns `string | number | symbol`.

Reply:
ah yes! great idea with that intersection :) I pushed out this change